py3status/modules/year_progress.py

```diff
@@ -0,0 +1,212 @@
+
+
+def compute_current_day_progress(current=None, end=None):
+    """Compute current hour progress to tomorrow (00:00 AM)"""
+    if not current:
+        current = datetime.now().astimezone()
+
+    # start at 00:00 AM
+    start = datetime(current.year, current.month, current.day).astimezone()
+    if not end:
+        end = datetime(
+            current.year, current.month, current.day, 23, 59, 00
+        ).astimezone()
+
+    return compute_progress(start, end, current)
+
+
+def compute_current_month_progress(current=None, end=None):
+    """Compute current date to the first date of  next month"""
+    if not current:
+        current = datetime.now(tz=UTC())
+
+    start = datetime(current.year, current.month, 1, tzinfo=UTC())
+    if not end:
+        # month max 12
+        if current.month == 12:
+            end = datetime(current.year, current.month, 31, tzinfo=UTC())
+        else:
+            end = datetime(current.year, current.month + 1, 1, tzinfo=UTC())
+    return compute_progress(start, end, current)
+
+
+def compute_current_year_progress(current=None, end=None):
+    """Compute current year progress to next year"""
+    if not current:
+        current = datetime.now(tz=UTC())
+
+    start = datetime(current.year, 1, 1, tzinfo=UTC())
+    if not end:
+        end = datetime(current.year + 1, 1, 1, tzinfo=UTC())
+    return compute_progress(start, end, current)
+
+
+class Py3status:
+    format = "{progress_bar} {ratio}%{mode}"
+    progress_width = 5
+    progress_block = "▓"
+    remain_block = "░"
+    cache_timeout = 3600
+    progress_mode = "year"
+    ctime = None
+    thresholds = 90  # precentage, .e.g 90%
+
+    def _check_urgency(self, thresholds, progress):
+        urgency = False
+        print(progress)
+        if progress > thresholds:
+            urgency = True
+
+        return urgency
+
+    def _create_progress_string(self, progress, width=20):
+        progress_int = int(round(progress * width))
+        rest_int = int(width - progress_int)
+        return "{}{}".format(
+            self.progress_block * progress_int, self.remain_block * rest_int
+        )
+
+    def year_progress(self):
+        end_time = None
+        if self.ctime:
+            end_time = datetime.strptime(self.ctime[0], self.ctime[1]).astimezone()
+
+        if self.progress_mode == "year":
+            progress_ratio = compute_current_year_progress(end=end_time)
+            mode = "y"
+        elif self.progress_mode == "month":
+            progress_ratio = compute_current_month_progress(end=end_time)
+            mode = "m"
+        elif self.progress_mode == "day":
+            progress_ratio = compute_current_day_progress(end=end_time)
+            mode = "d"
+        elif self.progress_mode == "week":
+            progress_ratio = compute_current_week_progress()
+            if end_time:  # can't use custom end time
+                progress_ratio = None
+            mode = "w"
```

Comment:
Kill `mode` in favor of `format = "[\?max_length=1 {progress_mode}]"`.

`{progress_mode}` should already work in `format` too due to the configs, but you can add `"progress_mode": self.progress_mode` to data too to make it more obvious.

Also, this is why we have `format`. It's for tricks too. :-)

Reply:
done.

Nice, thank you for pointing out.